src/test/util/filesystem-test-rig.ts

```diff
@@ -82,6 +82,38 @@ export class FilesystemTestRig {
     }
   }
 
+  /**
+   * Like {@link write}, but first writes to a temporary file, and then renames
+   * the file to its final location.
+   */
+  async writeAtomic(file: string, content: unknown): Promise<void>;
```

Comment:
oh nice, yeah, that could totally be causing some of the race conditions I was seeing on my PR. maybe this should be the default for write?

Reply:
Yeah this really confused me for a while! I never hit it locally, or in CI on Linux or macOS, but on Windows CI I would get "invalid JSON" errors at weird times on maybe 1/4 of the runs. I think the filesystem is fast enough everywhere else that it didn't manifest.

I think we're unlikely to hit this kind of problem in any tests apart from the watch mode tests, because everywhere else we wait for the write to complete before we invoke Wireit.

I updated all the write calls in the watch mode tests to use writeAtomic, though, since similar problems could theoretically occur when modifying input files in watch mode too.